src/github.com/getlantern/detour/detour.go

```diff
@@ -0,0 +1,353 @@
+			} else {
+				log.Debugf("Dial %s to %s failed: %s", dc.stateDesc(), addr, err)
+				return dc, err
+			}
+		}
+		// if whitelisted or dial directly failed, try detour
+		dc.setState(stateDetour)
+		dc.conn, err = dc.dialDetour(network, addr)
+		if err != nil {
+			log.Errorf("Dial %s to %s failed", dc.stateDesc(), addr)
+			return nil, err
+		}
+		log.Tracef("Dial %s to %s succeeded", dc.stateDesc(), addr)
+		if !whitelisted(addr) {
+			log.Tracef("Add %s to whitelist", addr)
+			addToWl(dc.addr, false)
+		}
+		return dc, err
+	}
+}
+
+// Read() implements the function from net.Conn
+func (dc *Conn) Read(b []byte) (n int, err error) {
+	if !dc.inState(stateInitial) {
+		return dc.followUpRead(b)
+	}
+	// state will always be settled after first read, safe to clear buffer at end of it
+	defer dc.resetLocalBuffer()
+	start := time.Now()
+	if !dc.readDeadline.IsZero() && dc.readDeadline.Sub(start) < 2*TimeoutToDetour {
+		log.Tracef("no time left to test %s, read %s", dc.addr, statesDesc[stateDirect])
+		dc.setState(stateDirect)
+		return dc.countedRead(b)
+	}
+	// wait for at most TimeoutToDetour to read
+	dc.getConn().SetReadDeadline(start.Add(TimeoutToDetour))
+	n, err = dc.countedRead(b)
+	dc.getConn().SetReadDeadline(dc.readDeadline)
+
+	detector := blockDetector.Load().(*Detector)
+	if err != nil {
+		log.Debugf("Error while read from %s %s: %s", dc.addr, dc.stateDesc(), err)
+		if detector.CheckError(err) {
+			return dc.detour(b)
+		}
+		return
+	}
+	if detector.CheckContent(b) {
+		log.Tracef("Read %d bytes from %s %s, content is hijacked, detour", n, dc.addr, dc.stateDesc())
+		return dc.detour(b)
+	}
+	log.Tracef("Read %d bytes from %s %s, set state to direct", n, dc.addr, dc.stateDesc())
+	dc.setState(stateDirect)
+	return
+}
+
+func (dc *Conn) followUpRead(b []byte) (n int, err error) {
+	detector := blockDetector.Load().(*Detector)
+	if n, err = dc.countedRead(b); err != nil {
+		if err == io.EOF {
+			log.Tracef("Read %d bytes from %s %s, EOF", n, dc.addr, dc.stateDesc())
+			return
+		}
+		log.Tracef("Read from %s %s failed: %s", dc.addr, dc.stateDesc(), err)
+		switch {
+		case dc.inState(stateDirect) && detector.CheckError(err):
+			log.Tracef("%s still blocked, add to whitelist so will try detour next time", dc.addr)
+			addToWl(dc.addr, false)
+		case dc.inState(stateDetour) && wlTemporarily(dc.addr):
+			log.Tracef("Detoured route is not reliable for %s, not whitelist it", dc.addr)
+			removeFromWl(dc.addr)
+		}
+		return
+	}
+	if dc.inState(stateDirect) && detector.CheckContent(b) {
+		log.Tracef("%s still content hijacked, add to whitelist so will try detour next time", dc.addr)
+		addToWl(dc.addr, false)
+		return
+	}
+	log.Tracef("Read %d bytes from %s %s", n, dc.addr, dc.stateDesc())
+	return
+}
+
+func (dc *Conn) detour(b []byte) (n int, err error) {
+	if err = dc.setupDetour(); err != nil {
+		log.Errorf("Error while setup detour: %s", err)
+		return
+	}
+	if _, err = dc.resend(); err != nil {
```

Comment:
Although it's not very likely, the worst case scenario here is a double form submission for a financial transaction, which would be very bad.  I think the safest thing to do is take it out.

At some point, I think we're going to want to have some detour-like logic that's aware of the HTTP protocol, to handle stuff like this and other potential things that are easier to detect at the HTTP level.